skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold
+
+import com.google.common.annotations.VisibleForTesting
+import com.intellij.openapi.components.ServiceManager
+import java.io.File
+import java.nio.file.Path
+import java.nio.file.Paths
+
+/**
+ * Interface for Skaffold execution service. This service builds and launches Skaffold process from
+ * the given set of standard flags and settings. [DefaultSkaffoldExecutorService] assumes Skaffold
+ * to be already installed and available.
+ */
+interface SkaffoldExecutorService {
+    companion object {
+        val instance
+            get() = ServiceManager.getService(SkaffoldExecutorService::class.java)!!
```

Comment:
Can we get rid of the `!!`?

Reply:
Also, do we need to call getService every time the property is accessed?

Reply:
This is a call to underlying IDE API, which could potentially return `null`, but that would indicate a configuration error that should never happen, hence `!!`. `getService()` is cached itself and the service can be changed dynamically, so we always call it as a getter. Instead of `!!`, do you recommend just declaring the `instance` as non-null? What would happen if `getService()` returns null?

Reply:
> Also, do we need to call getService every time the property is accessed?

So does that mean that even though its a `val` it will get called each time? I thought it would only get called once? @ivanporty something to look into

Reply:
or perhaps I misread the comment and @keyboardsurfer was just checking to make sure we didn't actually need to get it fresh each time.

Reply:
we will change that later a bit as well - when we have another implementation of executor available (the one using managed Skaffold installation), so we will choose implementation dynamically. So eventually I think it's a good idea to always get a current implementation using `ServiceManager`. Do we have any concerns about performance here?

Reply:
@ivanporty: Your service and potentially whole application would crash with an NPE if that'd return null.
[See doc](https://kotlinlang.org/docs/reference/null-safety.html#the--operator)

Calls to a custom getter are not cached. This means that your results can change over time, depending on the ServiceManager.

Reply:
Looking at this again, I think it makes sense to have this as a custom getter and defer the fetching of the service to the `ServiceManager` (which internally itself caches the service).

Regarding the `!!` and potential NPE, just an idea:

What if we created an extension function on `ServiceManager` that would check get the service and check if its null. If its null, then throw a new exception (something like `ApplicationServiceNotDefinedException` with a message). Its ok for this to bubble up and be an IDE exception in this case since we are the once defining this service, so it should really never be null (although we do still need to be very careful when loading across plugin.xml boundaries).

Reply:
oh right.... that sucks that you can't do this for static methods.

The separate function feels a lot less clean so I'm inclined to just leave it as is and let it blow up. After all, is this any different than the behavior in our other (java) plugin?

Reply:
Right, we would otherwise need to check for `null` all over the place. Missing our own service would indicate a serious error, so it would be same as in the Cloud Tools plugin.

Reply:
yeah. lets just leave it then. wdyt?